web/next.config.js

```diff
@@ -13,17 +12,61 @@ const withPWA = require('next-pwa')({
       handler: 'NetworkOnly',
     },
     {
-      urlPattern: (url) => { return url.pathname.startsWith("/admin/"); },
-      handler: 'NetworkOnly',
+      urlPattern: /\.(?:eot|otf|ttc|ttf|woff|woff2)$/i,
+      handler: 'StaleWhileRevalidate',
       options: {
-        fetchOptions: { credentials: 'same-origin' }
-      }
+        cacheName: 'static-font-assets',
+        expiration: {
+          maxEntries: 4,
+          maxAgeSeconds: 7 * 24 * 60 * 60, // 7 days
+        },
+      },
     },
     {
-      urlPattern: (url) => { return url.pathname.startsWith("/api/"); },
-      handler: 'NetworkOnly',
+      urlPattern: /\.(?:jpg|jpeg|gif|png|svg|ico|webp)$/i,
+      handler: 'StaleWhileRevalidate',
+      options: {
+        cacheName: 'static-image-assets',
+        expiration: {
+          maxEntries: 64,
+          maxAgeSeconds: 24 * 60 * 60, // 24 hours
+        },
+      },
+    },
+    {
+      urlPattern: /\.(?:mp4)$/i,
+      handler: 'CacheFirst',
+      options: {
+        rangeRequests: true,
+        cacheName: 'static-video-assets',
+        expiration: {
+          maxEntries: 32,
+          maxAgeSeconds: 24 * 60 * 60, // 24 hours
+        },
+      },
+    },
+    {
+      urlPattern: /\.(?:js)$/i,
+      handler: 'StaleWhileRevalidate',
+      options: {
+        cacheName: 'static-js-assets',
+        expiration: {
+          maxEntries: 32,
+          maxAgeSeconds: 24 * 60 * 60, // 24 hours
+        },
+      },
+    },
+    {
+      urlPattern: /\.(?:css)$/i,
+      handler: 'StaleWhileRevalidate',
+      options: {
+        cacheName: 'static-style-assets',
+        expiration: {
+          maxEntries: 32,
+          maxAgeSeconds: 24 * 60 * 60, // 24 hours
+        },
+      },
     },
```

Comment:
What could be the downsides here? I'm a little concerned about a scenario where the Owncast web resources are updated due to a new release, and some old items are pulled from the cache, and some new items are downloaded, causing an application exception. Same with CSS, a mix of old and new could break the UI. I don't know if this could happen or not.

We have HTTP cache header expirations already set on assets, so they should all get expired at the same time. But I don't know how this change would impact any of that.

What are the difference between these manual definitions and the previous `runtimeCaching` that was provided by the library?

From the docs:
> The [stale-while-revalidate](https://web.dev/articles/offline-cookbook#stale_while_revalidate) pattern allows you to respond to the request as quickly as possible with a cached response if available, falling back to the network request if it's not cached. The network request is then used to update the cache. As opposed to [some implementations](https://github.com/GoogleChrome/workbox/issues/2807) of stale-while-revalidate, this strategy will always make a revalidation request, regardless of the age of the cached response.
This is a fairly common strategy where having the most up-to-date resource is not vital to the application.

Reply:
Files that is compiled for production, nextjs will include `contenthash` as their name, shouldn't be a problem. If the files that placed inside public directory will be changed anytime, for example CSS files inside `public/styles/admin`, then I think there will be chances the admin UI will break.

The impact can be vary, from what I know, service worker will prevent to request network if the age specify in runtimeCaching is not expired. When runtimeCaching have shorter age than HTTP cache header, the request from service worker will be served by browser cache

About the different from previous `runtimeCaching`, from the library [cache.js](https://github.com/shadowwalker/next-pwa/blob/master/cache.js), the combination of last three config will match all apis and others routes, it is directly included to `runtimeCaching` which made all routes to run through service worker, causing `admin` route basic auth not showing on Firefox

Due to I lack of sufficient knowledge which Owncast assets is intended to be served by runtimeCaching, initial changes I only did is by removing `NetworkOnly` routes and copy exactly what written in library except last three config is removed, so the routes that is not explicitly stated will just skip service worker. I think if having the cache controlled by HTTP cache header is preferred, maybe we can just give empty array to runtimeCaching to prevent any unexpected caching happen on client side while still having the benefit of precache

Reply:
@gabek After some thought and testing, passing empty array to `runtimeCaching` seems to be a possible solution, `precaching` will working as usual.  When there is any assets is confirmed will benefit from `runtimeCaching` rather than only using HTTP cache header someday, we can then add config for that too

What do you think?

Reply:
That sounds like a good solution to me! It's mostly the precaching that this was all wired up for anyway, so not worrying about the runtime caching is probably the best bet. Thank you!